packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -143,13 +143,16 @@ export default class WorkspacesForeachCommand extends BaseCommand {
       ? rootCandidates.filter(fromPredicate)
       : rootCandidates;
 
-    const candidates = new Set([...fromCandidates, ...(await Promise.all(fromCandidates.map(async candidate => [...(
+    if (this.recursive && this.since)
```

Comment:
Hm, is what this implies true, namely, that `getRecursiveWorkspaceDependencies` is written in a way that does not require the install state? It appears so. That seems like a nice performance bump and to make `getRecursiveWorkspaceDependents` a little less fraught for anyone who might want to use it in the future. Would it be difficult to rewrite it as such?

Reply:
Thanks for the suggestion! It looks like it is possible, although it is slightly less efficient. Rather than just checking the set of dependencies for the workspace, we need to iterate through each dependency, check if it is a workspace, and then check if it is the workspace we are looking for. Its probably worth it though to avoid requiring the install state to be populated.

Fixed in d05f21e3a2556d8d35b775157511d52aa4eafad3.